instrumentation/kotlin-coroutines/src/main/java/io/opentelemetry/instrumentation/auto/kotlincoroutines/KotlinProbeInstrumentation.java

```diff
@@ -0,0 +1,193 @@
+        named("probeCoroutineSuspended").and(takesArguments(1)),
+        CoroutineSuspendedAdvice.class.getName());
+    return transformers;
+  }
+
+  @Override
+  public String[] helperClassNames() {
+    return new String[] {
+      "io.opentelemetry.instrumentation.auto.kotlincoroutines.KotlinProbeInstrumentation$CoroutineWrapper",
+      "io.opentelemetry.instrumentation.auto.kotlincoroutines.KotlinProbeInstrumentation$TraceScopeKey",
+      "io.opentelemetry.instrumentation.auto.kotlincoroutines.KotlinProbeInstrumentation$CoroutineContextWrapper",
+    };
+  }
+
+  public static class CoroutineCreatedAdvice {
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void exit(
+        @Advice.Return(readOnly = false) kotlin.coroutines.Continuation retVal) {
+      if (!(retVal instanceof CoroutineWrapper)) {
+        retVal = new CoroutineWrapper(retVal);
+      }
+    }
+  }
+
+  public static class CoroutineResumedAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(
+        @Advice.Argument(0) final kotlin.coroutines.Continuation continuation) {
+      continuation.getContext().get(TraceScopeKey.INSTANCE);
+    }
+  }
+
+  public static class CoroutineSuspendedAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(
+        @Advice.Argument(0) final kotlin.coroutines.Continuation continuation) {
+      continuation.getContext().minusKey(TraceScopeKey.INSTANCE);
+    }
+  }
+
+  public static class TraceScopeKey implements CoroutineContext.Key {
+    public static final TraceScopeKey INSTANCE = new TraceScopeKey();
+  }
+
+  public static class CoroutineWrapper implements kotlin.coroutines.Continuation {
+    private final Continuation proxy;
+    private final CoroutineContextWrapper contextWrapper;
+
+    public CoroutineWrapper(Continuation proxy) {
+      this.proxy = proxy;
+      this.contextWrapper = new CoroutineContextWrapper(proxy.getContext());
+    }
+
+    public String toString() {
+      return proxy.toString();
+    }
+
+    @NotNull
+    @Override
+    public CoroutineContext getContext() {
+      return contextWrapper;
+    }
+
+    @Override
+    public void resumeWith(@NotNull Object o) {
+      proxy.resumeWith(o);
+    }
+  }
+
+  public static class CoroutineContextWrapper implements CoroutineContext {
+    private final CoroutineContext proxy;
+    private Context myContext;
+    private Context prevContext;
+
+    public CoroutineContextWrapper(CoroutineContext proxy) {
+      this.proxy = proxy;
+      this.myContext = Context.current();
+    }
+
+    @Override
+    public <R> R fold(R r, @NotNull Function2<? super R, ? super Element, ? extends R> function2) {
+      return proxy.fold(r, function2);
+    }
+
+    @Nullable
+    @Override
+    public <E extends Element> E get(@NotNull Key<E> key) {
+      if (key == TraceScopeKey.INSTANCE) {
+        prevContext = myContext.attach();
```

Comment:
Are you sure that `get` with side effect is a good idea? `CoroutineResumedAdvice.enter` looks strange as well, when you `get` a value and then ignore it. May be you can extract context manipulation from `get` and `minusKey` and do that directly in advice methods?

Reply:
Good call; I found a cleaner way to do this (get() the context holder as its own value and call `tracingSuspend/tracingResume`) in a way that keeps the kotlin types happy.